workloads/Haskell/LuParser/src/Spec.hs

```diff
@@ -0,0 +1,101 @@
+{-# LANGUAGE ImportQualifiedPost #-}
+
+module Spec where
+
+import Data.Char (isAlphaNum, isLower, isSpace, isUpper)
+import Data.List (isInfixOf)
+import Etna.Lib
+import GHC.Unicode (isPrint)
+import Level
+import LuParser
+import LuSyntax
+import Parser (Parser)
+import Parser qualified as P
+import PrettyPrinter
+import Test.QuickCheck qualified as QC
+
+reserved :: [String]
+reserved =
+  [ "and",
+    "break",
+    "do",
+    "else",
+    "elseif",
+    "end",
+    "false",
+    "for",
+    "function",
+    "goto",
+    "if",
+    "in",
+    "local",
+    "nil",
+    "not",
+    "or",
+    "repeat",
+    "return",
+    "then",
+    "true",
+    "until",
+    "while"
+  ]
+
+doesNotContainEscapeChar :: String -> Bool
+doesNotContainEscapeChar s = not ("\"" `isInfixOf` s) && foldr (\x acc -> (isPrint x || isSpace x) && acc) True s
+
+doesNotContainReservedWords :: String -> Bool
+doesNotContainReservedWords s = not $ foldr (\x acc -> x `isInfixOf` s || acc) False Spec.reserved
+
+containsOnlyAlphaNumeralsOrUnderscore :: String -> Bool
+containsOnlyAlphaNumeralsOrUnderscore = foldr (\x acc -> (isAlphaNum x || x == '_') && acc) True
+
+isNotEmpty :: String -> Bool
+isNotEmpty s = s /= ""
+
+startWithLowerUpperOrUnderscore :: String -> Bool
+startWithLowerUpperOrUnderscore (x : _) = isUpper x || isLower x || x == '_'
+
+nameIsParsable :: VarName -> Bool
```

Comment:
Is this the only place where there's an actual precondition and the rest is just for recursing?

Reply:
These four are actual preconditions:
```
doesNotContainEscapeChar :: String -> Bool
doesNotContainEscapeChar s = not ("\"" `isInfixOf` s) && foldr (\x acc -> (isPrint x || isSpace x) && acc) True s

doesNotContainReservedWords :: String -> Bool
doesNotContainReservedWords s = not $ foldr (\x acc -> x `isInfixOf` s || acc) False Spec.reserved

containsOnlyAlphaNumeralsOrUnderscore :: String -> Bool
containsOnlyAlphaNumeralsOrUnderscore = foldr (\x acc -> (isAlphaNum x || x == '_') && acc) True

isNotEmpty :: String -> Bool
isNotEmpty s = s /= ""

startWithLowerUpperOrUnderscore :: String -> Bool
startWithLowerUpperOrUnderscore (x : _) = isUpper x || isLower x || x == '_'
```